src/Microsoft.ML.Transforms/GroupTransform.cs

```diff
@@ -180,35 +180,36 @@ public override RowCursor[] GetRowCursorSet(Func<int, bool> predicate, int n, Ra
         }
 
         /// <summary>
-        /// For group columns, the schema information is intact.
+        /// This class describes the relation between <see cref="GroupTransform"/>'s input <see cref="Schema"/>,
+        /// <see cref="GroupBinding._input"/>, and output <see cref="Schema"/>, <see cref="GroupBinding.OutputSchema"/>.
         ///
-        /// For keep columns, the type is Vector of original type and variable length.
-        /// The only metadata preserved is the KeyNames and IsNormalized.
+        /// The <see cref="GroupBinding.OutputSchema"/> contains columns used to group columns and columns being aggregated from input data.
+        /// In <see cref="GroupBinding.OutputSchema"/>, group columns are followed by aggregated columns. For example, if column "Age" is used to group "UserId" column,
+        /// the first column and the second column in <see cref="GroupBinding.OutputSchema"/> produced by <see cref="GroupTransform"/> would be "Age" and "UserId," respectively.
+        /// Note that "Age" is a group column while "UserId" is an aggregated (also call keep) column.
         ///
-        /// All other columns are dropped.
+        /// For group columns, the schema information is intact. For aggregated columns, the type is Vector of original type and variable length.
+        /// The only metadata preserved is the KeyNames and IsNormalized. All other columns are dropped. Please see
+        /// <see cref="GroupBinding.BuildOutputSchema(Schema)"/> how this idea got implemented.
         /// </summary>
-        private sealed class GroupSchema : ISchema
+        private sealed class GroupBinding
         {
-            private static readonly string[] _preservedMetadata =
-                new[] { MetadataUtils.Kinds.IsNormalized, MetadataUtils.Kinds.KeyValues };
-
             private readonly IExceptionContext _ectx;
             private readonly Schema _input;
 
+            // Column names in source schema used to group rows.
             private readonly string[] _groupColumns;
+            // Column names in source schema aggregated into row's vector-typed columns.
             private readonly string[] _keepColumns;
 
-            public readonly int[] GroupIds;
-            public readonly int[] KeepIds;
-
-            private readonly int _groupCount;
-            private readonly ColumnType[] _columnTypes;
-
-            private readonly Dictionary<string, int> _columnNameMap;
+            // GroupIds[i] is the i-th group(-key) column's column index in the source schema.
```

Comment:
>GroupIds [](start = 15, length = 8)

GroupIds looks to be renamed to GroupColumnIndexes - should this comment be updated?

Reply:
Right. Thank you.

---
In reply to: [245831781](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245831781) [](ancestors = 245831781)